### PR TITLE
thuang-517-AuthN

### LINKED ISSUE #571: Consumers want a matrix download to have a meaningful filename
For [Scientists can download annotated expression matrices from the portal dataset-view](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/16), there was agreement to use the **filename** for the download name. For this increment, the "-29-remixed" (-dsuid-remixed) must be eliminated from **filename**.

```
      "dataset_assets": [
        {
          "created_at": 1598479174.003882, 
          "dataset_id": "8e3bf114-5491-4855-99a1-83770a3f75fa", 
          "filename": "Single_cell_longitudinal_analysis_of_SARS_CoV_2_infection_in_human_bronchial_epithelial_cells-29-remixed.h5ad", 
          "filetype": "H5AD", 
```

---

Other options can be explored for customizing filenames based on consumer-defined formats and dataset metadata - [simple example](https://wiki.hydrogenaud.io/index.php?title=XLD_Configuration#.22File_Naming.22_tab).

### PR DESCRIPTION
**Pull request by tihuan**
When the user is logged in, you see the random scientist emoji:

Screen Shot 2020-11-12 at 2 06 03 PM

When a user is logged in and clicks on the emoji:
Screen Shot 2020-11-12 at 2 06 12 PM

UPDATE (11/16): Use email first char if name is not available

demo3
demo2

When user is not logged in, but they click on "Create Collection" button:
Screen Shot 2020-11-12 at 2 05 44 PM

(UPDATED: 11/13) When a user is not logged in and `auth` flag is `true`, they see "Log In" button:
Screen Shot 2020-11-13 at 11 20 42 AM

GIF:
demo

QA Steps:
1. Pull this branch: `thuang-517-AuthN`
1. Run `npm start`
1. Navigate to: `http://localhost:8000/?auth=true&cc=true` to enable both flags
1. You should now see "Log In" and "Sign Up" buttons, and if you click on "Create Collection" button, you'll get the CTA
1. Now to simulate a logged in user's experience, I use [Requestly](https://chrome.google.com/webstore/detail/requestly-redirect-url-mo/mdnleldcmiljblolnjhpnblkcekpdkpa?hl=en) to modify the request header's `Cookie` to include my `dev` env cookies. Essentially you log in on dev cellxgene (https://api.cellxgene.dev.single-cell.czi.technology/dp/v1/login), and then inspect a request we send to BE, so you can cookie the request header `Cookie` field and paste that into Requestly rule, with `contains URL: api.cellxgene.dev.single-cell.czi.technology`. Press "Save"

    E.g., 

    **NOTE: Please check your `frontend/src/configs/configs.js` uses `API_URL: "https://api.cellxgene.dev.single-cell.czi.technology"` instead of the old `corpora` namespace**

    Screen Shot 2020-11-12 at 2 29 35 PM

1. Now activate Requestly like so:
    
    Screen Shot 2020-11-12 at 2 34 30 PM

1. Refresh the page, and you should see all the logged in features!

1. Alternatively, you can also run local BE with `DEPLOYMENT_STAGE=test`! 

PTAL thanks all!

resolves #571

### CODE REVIEW

File: frontend/src/common/entities.ts [hunk 1]
```diff
@@ -27,9 +27,9 @@ export const COLLECTION_LINK_TYPE_OPTIONS = {
 };
 
 export interface Link {
-  name: string;
-  url: string;
-  type: COLLECTION_LINK_TYPE;
+  link_name: string;
```
Comment (tihuan) [inline frontend/src/common/entities.ts:30]: BE updated the keys, so changing them in the FE too 👌

File: frontend/src/common/featureFlags/features.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
 export enum FEATURES {
   CREATE_COLLECTION = "cc",
+  AUTH = "auth",
```
Comment (tihuan) [inline frontend/src/common/featureFlags/features.ts:3]: Add `auth` feature flag

File: frontend/src/common/queries/auth.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { useQuery } from "react-query";
```
Comment (tihuan) [inline frontend/src/common/queries/auth.ts:1]: Create `auth` related queries file

File: frontend/src/components/AppContainer/index.tsx [hunk 4]
```diff
@@ -21,9 +19,6 @@ const AsyncCollection = loadable<Props>(
   () => /*webpackChunkName: 'AsyncCollection' */ import("src/views/Collection")
 );
 
-// (thuang): Check for Create Collection feature flag
-get(FEATURES.CREATE_COLLECTION);
```
Comment (tihuan) [inline frontend/src/components/AppContainer/index.tsx:25]: For modularization, `get(FEATURES.CREATE_COLLECTION)` is used at the callsite (components that care about this flag), so we don't need centralized management for feature flags, which causes problems such as not knowing if deleting this call will break any components

File: frontend/src/components/CreateCollectionModal/index.tsx [hunk 5]
```diff
@@ -13,31 +14,60 @@ const AsyncContent = loadable(
     )
 );
 
+const AsyncCTA = loadable(
```
Comment (tihuan) [inline frontend/src/components/CreateCollectionModal/index.tsx:17]: Code split and lazy load CTA only when the condition is met, so we don't send down unnecessary bytes to users

**Comment by tihuan** [comment]
Thanks so much for linking this PR to #517, @brianraymor ! 🙏  I'll remember to add the Zenhub ticket in PR description next time!

**Comment by tihuan** [comment]
Just updated the PR to only have "Log In" button and removed "Sign Up" button 🙏 PTAL again! Thank you!

**Review by hthomas-czi — COMMENTED** [review]
I thought we were able to get the user's first name from Auth0 and could display that?

**Comment by Bento007** [comment]
What is the point of `continue` when you click collections when not logged in?

File: frontend/src/components/CreateCollectionModal/index.tsx [hunk 6]
```diff
@@ -13,31 +14,60 @@ const AsyncContent = loadable(
...
+        content: AsyncCTA,
+        isCloseButtonShown: true,
+        title: "Create an account or sign-in to get started",
+      };
 
   return (
     <>
       <StyledButton
-        onMouseOver={() => AsyncContent.preload()}
+        onMouseOver={() => config.content.preload()}
```
Comment (seve) [inline frontend/src/components/CreateCollectionModal/index.tsx:50]: So cool!

**Review by seve — APPROVED** [review]
LGTM!

**Comment by tihuan** [comment]
@Bento007 great question! So when a user clicks on "Continue", it takes them to the login/signup page!

@hthomas-czi thanks for catching that! Just updated the code to show user's name when available 💪 

Screen Shot 2020-11-16 at 12 19 56 PM

**Review by Bento007 — APPROVED** [review]
LGTM

**Comment by tihuan** [comment]
@hthomas-czi Updated to use first char of email if user's name is an email!

demo3

The only issue I see is on hover, the circle will disappear, since it's the same color (Blue 6) as the background!
demo2

Thank you!

**Review by hthomas-czi — APPROVED** [review]
Beautiful! Great work, Timmy.

**Comment by tihuan** [comment]
Whoop whoop! Always enjoy our collaboration, @hthomas-czi ! And thanks @seve @Bento007 for the swift reviews too 🚀 ✨ MERGING!

**Comment by tihuan** [comment]
Just merged this in `main`! Not in Dev env yet!

**Pull request merged**